### ISSUE TITLE
[feat] Suggestion on better interface

### ISSUE THREAD

**Issue by LamprosPitsillos** [ziontee113/SnippetGenie#1]
![genie_15-02-2023_W](https://user-images.githubusercontent.com/61395246/219012129-274be9b4-07a2-4c77-8477-4417b97896de.png)

By selecting the function the node `local_declaration: function_declaration [2, 0] - [5, 3]` is selected , and the default behavior could be to add insert nodes on the `name`s and `body` of the parent ,selected, node.

**Comment by LamprosPitsillos** [comment]
Maybe have no default behaviour,but still leverage treesitter to select nodes instead of using regex and visual selection

**Comment by ziontee113** [comment]
Hi @LamprosPitsillos, to reiterate, are you suggesting after selecting the `function_declaration` node, the user won't have to select other nodes manually, but rather have them automatically turned into insert nodes?

I did try that approach and decided to not pursue that path, with couple of reasons:
1. Scalability: when it comes to larger, more complicated snippets, the snippet will be very large and bloated. Littered with insert nodes that the user might or might not want. Which kind of nodes being automatically selected can be limited with Tree-sitter queries, but still, there's no fine grain control over the selections.
2. After some research, I'm thinking about a separate Neovim plugin that would allow the user to jump / select nodes base on the queries. This will give the user more fine-tuned control over which part of the snippet they want to turn into insert nodes.

If you have other ideas, please let me know. Thank you.

**Comment by LamprosPitsillos** [comment]
How about this idea ? 
> Maybe have no default behaviour,but still leverage treesitter to select nodes instead of using regex and visual selection

Now i use `<CR>`  to select a node in the treesitter tree 
```
            incremental_selection = {
                enable = true,
                keymaps = {
                    init_selection = "<CR>",
                    node_incremental = "<CR>",
                    scope_incremental = "<Tab>",
                    node_decremental = "<S-Tab>",
                },
            },

```
What if you select the parent node first , go to a `register insert node mode` , and start selecting and registering nodes ,and when you are done the snippet is registered.
Of course this would have to be separate from visually selecting `insert nodes`
an example : 

https://user-images.githubusercontent.com/61395246/219030665-59a064ef-defd-4a97-bf04-89abc2e859fe.mp4

A counter argument to this would be that using TS incremental selection separate to `SnippetGenie` and just selecting it using the 
TS incremental selection is just the same without any changes to the way the plugin works , this could maybe be mentioned in the ReadMe as a suggested companion plugin ( [ts textobjects](https://github.com/nvim-treesitter/nvim-treesitter-textobjects) ) to this.

Maybe a more unique feature could be to say which TS nodes you would want to extract for a snippet and just select it
and the nodes get extracted automatically  ( but as you said you discarded the automation features ) but none the less i would give an example : 
` initiate the selection -> say which nodes to get extracted ( name ,body ) -> name the snip maybe - >done`

This is all just a fun discussion and would love to chat more about it : )

**Comment by ziontee113** [comment]
> What if you select the parent node first , go to a register insert node mode , and start selecting and registering nodes ,and when you are done the snippet is registered.
Of course this would have to be separate from visually selecting insert nodes

I assume you're proposing that instead of using a hotkey like `<CR>`, the user would enter visual mode, make a selection, and when they press `<Esc>` to go back to normal mode, an insert node is registered?

This approach would be inconvenient for the user when they make mistakes. Besides, you can always use a different key rather than `<CR>`.

> Maybe a more unique feature could be to say which TS nodes you would want to extract for a snippet and just select it
and the nodes get extracted automatically ( but as you said you discarded the automation features ) but none the less i would give an example :
 initiate the selection -> say which nodes to get extracted ( name ,body ) -> name the snip maybe - >done

This approach sounds promising, although it would be costly to implement. To implement this feature first I'd have to implement a UI first, which shows the user the state of the current snippet creation. This feature you mention would be an extension of the UI, where the user can press buttons to toggle which type of TS node they want to  extract.

**Comment by LamprosPitsillos** [comment]
Yes and no, i don't think it _needs_ to be this complected , but yes if the feature was to be polished then yes.
The simplest solution to this i can think is with  `vim.ui.select()` :

```lua
-- from the docs :h vim.ui.select
     vim.ui.select({ --[[ fill this with TSnodes in selected part ]] }, {
         prompt = 'Select nodes from selection:',
         format_item = function(item)
             return "I'd like to choose " .. item
         end,
     }, function(choice)
        --[[ code to add choice to selected nodes ,
recall this function ( minus previously selected nodes ) until input is nil/empty
          ]]
     end)
```
I am not a lua expert so forgive me if i oversimplify stuff, but this sound possible

**Comment by ziontee113** [comment]
If I understand correctly, you're suggesting using `vim.ui.select` to the let user pick which node they'd like to turn into an Insert Node. 

If that's what you meant, that's a cool idea. However things will get complicated when we encounter nodes that contains other nodes. Also I don't think `vim.ui.select` is a good long term solution.

It's a cool idea, but will need significant work to implement and evolve.
